### PR TITLE
Adds code generation via `@Encodable` annotation.

### PR DESCRIPTION
**Pull request by vkryachko**
Supported features:

* Codegen is based on public java-bean style getters, fields are not encoded
  directly
* Full support for generics
* Supports `java.util.Optional` via `value.getOrElse(null)`
* Fields can be marked as `@Encodable.Ignore` to skip them during encoding
* Fields can be marked as `@Encodable.Field(name="foo")` for custom encoded name

### CODE REVIEW

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/Names.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  /**
+   * Returns the class name that corresponds to a given element.
+   *
+   * <p>For top level classes returns the class name unchanged. For nested classes returns {@code
+   * ParentNested} for {@code Parent$Nested} classes.
+   */
+  static String generatedClassName(Element element) {
+    StringBuilder sb = new StringBuilder(element.getSimpleName().toString());
+    Element enclosingElement = element.getEnclosingElement();
+    while (!(enclosingElement instanceof PackageElement)) {
```
Comment (rlazo) [inline encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/Names.java]: Could you add a separator like AutoValue's `_`?

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/VisitResult.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import androidx.annotation.Nullable;
+import com.google.auto.value.AutoValue;
+import java.util.Collections;
+import java.util.Set;
+import javax.lang.model.type.TypeMirror;
+
+@AutoValue
+abstract class VisitResult<T> {
+  abstract Set<TypeMirror> moreToVisit();
```
Comment (rlazo) [inline encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/VisitResult.java]: what about `pendingToVisit`

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    TypeMirror map = types.erasure(elements.getTypeElement("java.util.Map").asType());
+    if (types.isAssignable(returnType, map)) {
+      TypeMirror keyType = ((DeclaredType) returnType).getTypeArguments().get(0);
+      if (!types.isSameType(keyType, elements.getTypeElement("java.lang.String").asType())) {
+        messager.printMessage(
+            Diagnostic.Kind.ERROR,
+            "Cannot encode Maps with non-String keys.",
+            ((DeclaredType) returnType).asElement());
+      }
+    }
```
Comment (rlazo) [inline encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java]: Wouldn't it be better to leave this on the encoders? If we ever support more than String for keys, we would have to remember to change 2 places.

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    if (types.isAssignable(
+        returnType, types.erasure(elements.getTypeElement("java.util.Optional").asType()))) {
+      returnType = ((DeclaredType) returnType).getTypeArguments().get(0);
+      getterExpression = getterExpression + ".orElse(null)";
+    }
```
Comment (rlazo) [inline encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java]: How likely is that we need to support optional? I though it wasn't available in android in anything less than API24

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+    return Optional.of(Getter.create(fieldName.get(), getterExpression, returnType));
+  }
+
+  // ownerType = Bar<Optional<Integer>, Boolean>
+  // ownerElement = Bar<T, U>
+  // genericType = U
+  private TypeMirror resolveTypeArguments(DeclaredType ownerType, TypeMirror genericType) {
```
Comment (rlazo) [inline encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java]: comment?

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        && methodName.length() != 2
+        && method.getReturnType().getKind() == TypeKind.BOOLEAN) {
+      return Optional.of(Character.toLowerCase(methodName.charAt(2)) + methodName.substring(3));
+    }
+
+    if (methodName.startsWith("get")
+        && methodName.length() != 3
+        && method.getReturnType().getKind() != TypeKind.VOID) {
+      return Optional.of(Character.toLowerCase(methodName.charAt(3)) + methodName.substring(4));
+    }
```
Comment (rlazo) [inline encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java]: This is perhaps something good to note in the comments.

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/EncodableProcessor.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    Set<Encoder> encoders = TypeTraversal.traverse(element.asType(), new GetterVisitor());
+
+    MethodSpec.Builder configureMethod =
+        MethodSpec.methodBuilder("configure")
+            .addParameter(
+                ParameterizedTypeName.get(
+                    ClassName.get("com.google.firebase.encoders.config", "EncoderConfig"),
+                    WildcardTypeName.subtypeOf(Object.class)),
+                "cfg")
+            .addModifiers(Modifier.PUBLIC)
```
[the remark this answers is not in the document]
Comment (vkryachko) [reply encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/EncodableProcessor.java]: afaict the lint runs on source only, but added a hide for completeness.

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/Names.java (hunk 1 above)
Comment (vkryachko) [reply encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/Names.java]: I am not a fan of `_`, also fwiw dagger2 does not use them either, so unless you feel strongly I'd prefer not to have them.

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java (hunk 3 above)
Comment (vkryachko) [reply encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java]: I think until we do, it's better DevX to fail at compile time than at runtime, when/if we add support for non-strings we can always just remove this code, wdyt?

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java (hunk 4 above)
Comment (vkryachko) [reply encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java]: It's available in google3 and will be supported by agp 4 desugaring.

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java (hunk 5 above)
Comment (vkryachko) [reply encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java]: removed, just notes I used when implementing this thing :)

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java (hunk 6 above)
Comment (vkryachko) [reply encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java]: Good point, updated `@Encodable`s javadoc  to address it.

File: encoders/firebase-encoders-json/src/main/java/com/google/firebase/encoders/annotations/Encodable.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ * Indicates to automatically generate an encoder capable to encode a given type.
+ *
+ * <p>All fields that need to be encoded <strong>must</strong>must have Java-bean compliant getters,
+ * otherwise they will not be encoded.
+ *
+ * Using the generated encoder:
+ * <pre>{@code
+ * @Encodable
+ * public class MyType {
+ *   public String getField() { return "hello"; }
```
Comment (rlazo) [inline encoders/firebase-encoders-json/src/main/java/com/google/firebase/encoders/annotations/Encodable.java]:
```suggestion
 *   public String getField() { return "hello"; }
 *   public boolean isConditional() { return false; }
```

Comment (rlazo) [reply encoders/firebase-encoders-json/src/main/java/com/google/firebase/encoders/annotations/Encodable.java]: Just for completeness.

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java (hunk 3 above)
Comment (rlazo) [reply encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java]: Sure, let's add a comment to point that out then.

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java (hunk 4 above)
Comment (rlazo) [reply encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java]: lgtm

File: encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java (hunk 6 above)
Comment (rlazo) [reply encoders/firebase-encoders-processor/src/main/java/com/google/firebase/encoders/processor/getters/GetterFactory.java]: Thanks!

File: root-project.gradle [hunk 9]
```diff
@@ -80,15 +80,15 @@ configure(subprojects) {
     }
     tasks.googleJavaFormat {
         source '.'
-        include '**/*.java'
+        include '**/java/**/*.java'
```
Comment (rlazo) [inline root-project.gradle]: why is this needed?

Comment (rlazo) [reply root-project.gradle]: nvm I saw the changes in the other PR
